src/SimpleSMT/Solver/Process.hs

```diff
@@ -95,13 +105,29 @@ with exe args logger todo = do
 toBackend :: SolverProcess -> Solver.Backend
 toBackend solver =
   Solver.Backend $ \cmd -> do
-    LBS.hPutStrLn (getStdin $ process solver) cmd
-    toLazyByteString <$> (hGetContentsEager mempty $ getStdout $ process solver)
+    hPutBuilder (getStdin $ process solver) cmd
+    getNextOutput $ response solver
   where
-    hGetContentsEager :: Builder -> Handle -> IO Builder
-    hGetContentsEager acc h = do
-      chunk <- LBS.hGet h 1024
-      let acc' = acc <> lazyByteString chunk
-      if chunk == mempty
-        then return acc'
-        else hGetContentsEager acc' h
+    getNextOutput :: IORef LBS.ByteString -> IO LBS.ByteString
+    getNextOutput solverResponse = atomicModifyIORef solverResponse $ \buffer ->
```

Comment:
I'm wondering, is using lazy IO necessary? The `send` function of the backend could read lines with hGetLine and feed them to the parser until it succeeds.

Reply:
The parsing here looks similar in purpose to
```
-- | Returns 0 when parentheses are balanced
-- 
-- >  scanParen 0 "(()())))" == -2
-- >  scanParen 0 "(()" == 1
--
scanParen :: Int -> String -> Int
scanParen s ('(':xs) = scanParen (s+1) xs
scanParen s (')':xs) = scanParen (s-1) xs
scanParen s [] = s
scanParen s (_:xs) = scanParen s xs
```

This would need to be expressed for bytestrings, but do we need more to detect if an expression is "complete"?

Reply:
Lazy I/O is more versatile (there might be solvers out there which don't break lines between outputs) but the reason I was using it here is that it's simpler to use than repeatedly calling `hGetLine`.

But yeah I think I'll switch to using `hGetLine` because it is blocking hence ensures that we did get the full output (with the current solution it might be the case that we detect an incomplete output hence return the whole output just because the solver isn't done).

Yes `scanParen` is what I'm doing here. But I don't think it is enough: while the solver shouldn't be outputting comments, I think it may output an error message such as `(wrong character on line 1 column 34, '(')`.